graphstorm-processing/graphstorm_processing/data_transformations/dist_transformations/dist_hard_negative_transformation.py

```diff
@@ -0,0 +1,123 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+from typing import Sequence
+from pyspark.sql.functions import split, col
+from pyspark.sql.types import ArrayType, IntegerType, StringType
+from pyspark.sql import DataFrame, functions as F, SparkSession
+
+from graphstorm_processing.constants import (
+    NODE_MAPPING_STR,
+    NODE_MAPPING_INT,
+    ORDER_INDEX,
+    EXPLODE_HARD_NEGATIVE_VALUE,
+)
+
+from .base_dist_transformation import DistributedTransformation
+
+
+class DistHardEdgeNegativeTransformation(DistributedTransformation):
+    """Transformation to apply hard negative transformation.
+
+    Parameters
+    ----------
+    cols : Sequence[str]
+        List of column names to apply hard negative transformation to.
+    spark: SparkSession
+        The spark session.
+    hard_node_mapping_dict: dict
+        The mapping dictionary contain mapping file directory and edge type.
+        {
+            "edge_type": str
+                Edge type to apply hard negative transformation.
+            "mapping_path": str
+                Path to the raw node mapping.
+            "format_name": str
+                Parquet.
+        }
+    separator: str, optional
+        The separator for string input value. Only required when input value type is string.
+    """
+
+    def __init__(
+        self,
+        cols: Sequence[str],
+        spark: SparkSession,
+        hard_node_mapping_dict: dict,
+        separator: str = "",
+    ) -> None:
+        super().__init__(cols, spark)
+        self.cols = cols
+        assert len(self.cols) == 1, "Hard Negative Transformation only supports single column"
+        self.separator = separator
+        self.hard_node_mapping_dict = hard_node_mapping_dict
+        assert self.hard_node_mapping_dict, "edge mapping dict cannot be None for hard negative "
+
+    def apply(self, input_df: DataFrame) -> DataFrame:
+        assert self.spark
+        input_col = self.cols[0]
+        column_type = input_df.schema[input_col].dataType
+        if isinstance(column_type, StringType):
+            transformed_df = input_df.withColumn(input_col, split(col(input_col), self.separator))
+        else:
+            transformed_df = input_df
+        # Edge type should be (src_ntype:relation_type:dst_ntype)
+        # Only support hard negative for destination nodes. Get the node type of destination nodes.
+        # TODO: support hard negative for source nodes.
+        _, _, dst_type = self.hard_node_mapping_dict["edge_type"].split(":")
+        mapping_prefix = self.hard_node_mapping_dict["mapping_path"]
+        format_name = self.hard_node_mapping_dict["format_name"]
+        hard_negative_node_mapping = self.spark.read.parquet(
+            f"{mapping_prefix}{dst_type}/{format_name}/"
+        )
+        max_size = (
```

Comment:
Add a comment here explaining what max_size (the maximum number of negatives in the input data)